### PR TITLE
refactor: Register and fetch webhook url for dedicated agent

### PR DESCRIPTION
**Pull request by bhavanakarwade**
### What?
- Register webhook url for dedicated agent
- Fetch webhook url for dedicated agent
- Support datetime-local datatype while schema creation

### CODE REVIEW

File: apps/api-gateway/src/issuance/issuance.controller.ts [hunk 1]
```diff
@@ -723,13 +723,15 @@ issueCredentialDto.type = 'Issuance';
...
-      const  webhookUrl = await this.issueCredentialService._getWebhookUrl(issueCredentialDto.contextCorrelationId).catch(error => {
+      };   
+      
+      const  webhookUrl = await this.issueCredentialService._getWebhookUrl(issueCredentialDto.contextCorrelationId, id).catch(error => {
         this.logger.debug(`error in getting webhook url ::: ${JSON.stringify(error)}`);
       });      
       if (webhookUrl) {
-        
-          await this.issueCredentialService._postWebhookResponse(webhookUrl, {data:issueCredentialDto}).catch(error => {
+        const plainIssuanceDto = JSON.parse(JSON.stringify(issueCredentialDto));
```
Comment (pallavighule) [inline apps/api-gateway/src/issuance/issuance.controller.ts:732]: Any reason to parsing data here?

File: apps/api-gateway/src/connection/connection.service.ts [hunk 2]
```diff
@@ -95,9 +95,12 @@ export class ConnectionService extends BaseService {
...
   }
 
-  async _getWebhookUrl(tenantId: string): Promise<string> {
+  async _getWebhookUrl(tenantId?: string, orgId?: string): Promise<string> {
     const pattern = { cmd: 'get-webhookurl' };
-    const payload = { tenantId };
+
+    const getWebhook = { tenantId, orgId };
+    
+    const payload = { getWebhook };
```
Comment (tipusinghaw) [inline apps/api-gateway/src/connection/connection.service.ts:103]:
```suggestion
    const payload = { tenantId, orgId};
```

File: apps/api-gateway/src/issuance/issuance.service.ts [hunk 3]
```diff
@@ -129,9 +129,9 @@ export class IssuanceService extends BaseService {
         return this.sendNatsMessage(this.issuanceProxy, 'retry-bulk-credentials', payload);
     }
 
-    async _getWebhookUrl(tenantId: string): Promise<string> {
+    async _getWebhookUrl(tenantId?: string, orgId?: string): Promise<string> {
```
Comment (tipusinghaw) [inline apps/api-gateway/src/issuance/issuance.service.ts:132]: `tenantId`,`orgId` both are optional parameter to get webhook URL ?

File: apps/issuance/interfaces/issuance.interfaces.ts [hunk 4]
```diff
@@ -141,6 +141,15 @@ export interface ICredential{
   '@context':[];
   type: string[];
   prettyVc?: IPrettyVc;
+  issuer?: {
```
Comment (tipusinghaw) [inline apps/issuance/interfaces/issuance.interfaces.ts:144]: Any specific reason to use `issuer` as optional parameter ?

**Review by pallavighule — APPROVED** [review]
LGTM

File: apps/webhook/src/webhook.repository.ts [hunk 5]
```diff
@@ -28,14 +29,26 @@ export class WebhookRepository {
...
-        where: {
-          tenantId
-        }
-      });
 
+      const { tenantId, orgId } = getWebhook;
+      let webhookUrlInfo;
+
+      if ('default' === tenantId && orgId) {
+        webhookUrlInfo = this.prisma.org_agents.findFirst({
```
Comment (tipusinghaw) [inline apps/webhook/src/webhook.repository.ts:39]:
```suggestion
        webhookUrlInfo = this.prisma.org_agents.findFirstOrThrow({
```

File: apps/api-gateway/src/issuance/issuance.controller.ts (hunk 1 above)
Comment (bhavanakarwade) [reply apps/api-gateway/src/issuance/issuance.controller.ts:732]: The data is being parsed to make a copy, so the original issueCredentialDto stays unchanged when sending it to the webhook.

**Review by tipusinghaw — APPROVED** [review]

File: apps/api-gateway/src/issuance/issuance.service.ts (hunk 3 above)
Comment (bhavanakarwade) [reply apps/api-gateway/src/issuance/issuance.service.ts:132]: Yes, for shared agent, we need tenantId for fetching the webhook Url details and for dedicated agent we need OrgId for fetching the details, That's why I have made both the arguments are optional

File: apps/webhook/src/webhook.repository.ts (hunk 5 above)
Comment (bhavanakarwade) [reply apps/webhook/src/webhook.repository.ts:39]: Resolved

File: apps/issuance/interfaces/issuance.interfaces.ts (hunk 4 above)
Comment (bhavanakarwade) [reply apps/issuance/interfaces/issuance.interfaces.ts:144]: Yes, we have used that interface at many places, So to ensure that the previous changes work correctly, it has been made optional

File: apps/api-gateway/src/connection/connection.service.ts (hunk 2 above)
Comment (bhavanakarwade) [reply apps/api-gateway/src/connection/connection.service.ts:103]: This function is calling the webhook controller, and in the webhook service controller, we use it in that format while listening. That's why I’ve bound the tenantId and orgId to the payload and sent it

**Review by pallavighule — APPROVED** [review]

**Pull request merged**